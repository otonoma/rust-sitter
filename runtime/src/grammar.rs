//! Grammar related functions.
use std::collections::{HashMap, HashSet};

use serde::{Deserialize, Serialize};

// NOTE: This could be useful for generating the grammar in the first place instead of just
// building json! values directly.

/// Type for the JSON representation of a grammar, mostly copied from `tree_sitter_generate`.
#[derive(Deserialize, Serialize)]
pub struct Grammar {
    pub name: String,
    pub word: Option<String>,
    // NOTE: Use `indexmap` because we need to preserve order.
    // https://docs.rs/indexmap/2.10.0/indexmap/map/struct.IndexMap.html
    pub rules: HashMap<String, RuleDef>,
    pub extras: Vec<RuleDef>,
}

#[derive(Deserialize, Serialize)]
#[serde(tag = "type")]
#[allow(non_camel_case_types)]
#[allow(clippy::upper_case_acronyms)]
pub enum RuleDef {
    ALIAS {
        content: Box<RuleDef>,
        named: bool,
        value: String,
    },
    BLANK,
    STRING {
        value: String,
    },
    PATTERN {
        value: String,
        flags: Option<String>,
    },
    SYMBOL {
        name: String,
    },
    CHOICE {
        members: Vec<RuleDef>,
    },
    FIELD {
        name: String,
        content: Box<RuleDef>,
    },
    SEQ {
        members: Vec<RuleDef>,
    },
    REPEAT {
        content: Box<RuleDef>,
    },
    REPEAT1 {
        content: Box<RuleDef>,
    },
    PREC_DYNAMIC {
        value: i32,
        content: Box<RuleDef>,
    },
    PREC_LEFT {
        value: PrecedenceValue,
        content: Box<RuleDef>,
    },
    PREC_RIGHT {
        value: PrecedenceValue,
        content: Box<RuleDef>,
    },
    PREC {
        value: PrecedenceValue,
        content: Box<RuleDef>,
    },
    TOKEN {
        content: Box<RuleDef>,
    },
    IMMEDIATE_TOKEN {
        content: Box<RuleDef>,
    },
    RESERVED {
        context_name: String,
        content: Box<RuleDef>,
    },
}

#[derive(Deserialize, Serialize)]
#[serde(untagged)]
pub enum PrecedenceValue {
    Integer(i32),
    Name(String),
}

impl Grammar {
    /// Starting from `rule_name`, find all symbols (named or anonymous) which can be reached.
    pub fn reachable_set<'a>(&'a self, rule_name: &str) -> Option<HashSet<&'a str>> {
        let mut set = HashSet::new();
        let (name, rule) = self.rules.get_key_value(rule_name)?;
        set.insert(name.as_str());
        self.compute_reachable(rule, &mut set)?;
        Some(set)
    }

    fn compute_reachable<'a>(
        &'a self,
        rule: &'a RuleDef,
        set: &mut HashSet<&'a str>,
    ) -> Option<()> {
        match rule {
            RuleDef::ALIAS {
                content,
                named: _,
                value,
            } => {
                if set.insert(value) {
                    self.compute_reachable(content, set)?;
                }
            }
            RuleDef::BLANK => {}
            RuleDef::STRING { value } => {
                set.insert(value.as_str());
            }
            RuleDef::PATTERN { value: _, flags: _ } => {}
            RuleDef::SYMBOL { name } => {
                // Don't repeat if we have already seen it before.
                if set.insert(name.as_str()) {
                    let rule = self.rules.get(name)?;
                    self.compute_reachable(rule, set)?;
                }
            }
            RuleDef::CHOICE { members } => {
                for member in members {
                    self.compute_reachable(member, set)?;
                }
            }
            RuleDef::FIELD { name: _, content } => self.compute_reachable(content, set)?,
            RuleDef::SEQ { members } => {
                for member in members {
                    self.compute_reachable(member, set)?;
                }
            }
            RuleDef::REPEAT { content } => self.compute_reachable(content, set)?,
            RuleDef::REPEAT1 { content } => self.compute_reachable(content, set)?,
            RuleDef::PREC_DYNAMIC { value: _, content } => self.compute_reachable(content, set)?,
            RuleDef::PREC_LEFT { value: _, content } => self.compute_reachable(content, set)?,
            RuleDef::PREC_RIGHT { value: _, content } => self.compute_reachable(content, set)?,
            RuleDef::PREC { value: _, content } => self.compute_reachable(content, set)?,
            RuleDef::TOKEN { content } => self.compute_reachable(content, set)?,
            RuleDef::IMMEDIATE_TOKEN { content } => self.compute_reachable(content, set)?,
            RuleDef::RESERVED {
                context_name: _,
                content,
            } => self.compute_reachable(content, set)?,
        }

        Some(())
    }
}
