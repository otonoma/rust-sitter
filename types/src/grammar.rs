//! Grammar related functions.
use std::collections::HashSet;

use indexmap::IndexMap;
use serde::{Deserialize, Serialize};

// NOTE: This could be useful for generating the grammar in the first place instead of just
// building json! values directly.

/// Type for the JSON representation of a grammar, mostly copied from `tree_sitter_generate`.
#[derive(Deserialize, Serialize)]
pub struct Grammar {
    pub name: String,
    pub word: Option<String>,
    pub rules: IndexMap<String, RuleDef>,
    pub extras: Vec<RuleDef>,
}

#[derive(Clone, Debug, PartialEq, Eq, Deserialize, Serialize)]
#[serde(tag = "type")]
#[allow(non_camel_case_types)]
#[allow(clippy::upper_case_acronyms)]
pub enum RuleDef {
    ALIAS {
        content: Box<RuleDef>,
        named: bool,
        value: String,
    },
    BLANK,
    STRING {
        value: String,
    },
    PATTERN {
        value: String,
        #[serde(skip_serializing_if = "Option::is_none")]
        flags: Option<String>,
    },
    SYMBOL {
        name: String,
    },
    CHOICE {
        members: Vec<RuleDef>,
    },
    FIELD {
        name: String,
        content: Box<RuleDef>,
    },
    SEQ {
        members: Vec<RuleDef>,
    },
    REPEAT {
        content: Box<RuleDef>,
    },
    REPEAT1 {
        content: Box<RuleDef>,
    },
    PREC_DYNAMIC {
        value: i32,
        content: Box<RuleDef>,
    },
    PREC_LEFT {
        value: PrecedenceValue,
        content: Box<RuleDef>,
    },
    PREC_RIGHT {
        value: PrecedenceValue,
        content: Box<RuleDef>,
    },
    PREC {
        value: PrecedenceValue,
        content: Box<RuleDef>,
    },
    TOKEN {
        content: Box<RuleDef>,
    },
    IMMEDIATE_TOKEN {
        content: Box<RuleDef>,
    },
    RESERVED {
        context_name: String,
        content: Box<RuleDef>,
    },
}

impl RuleDef {
    pub fn is_symbol(&self) -> bool {
        matches!(self, RuleDef::SYMBOL { .. })
    }

    pub fn is_blank(&self) -> bool {
        matches!(self, RuleDef::BLANK)
    }

    pub fn optional(rule: RuleDef) -> RuleDef {
        RuleDef::CHOICE {
            members: vec![RuleDef::BLANK, rule],
        }
    }

    pub fn as_optional(&self) -> Option<&RuleDef> {
        match self {
            Self::CHOICE { members } => match members.as_slice() {
                &[ref rule, RuleDef::BLANK] | &[RuleDef::BLANK, ref rule] => Some(rule),
                _ => None,
            },
            Self::PREC { value: _, content }
            | Self::PREC_LEFT { value: _, content }
            | Self::PREC_RIGHT { value: _, content }
            | Self::PREC_DYNAMIC { value: _, content } => content.as_optional(),
            _ => None,
        }
    }

    /// Pull out a sequence, including through precedence unwrapping.
    pub fn as_seq(&self) -> Option<&[RuleDef]> {
        match self {
            Self::SEQ { members } => Some(members),
            Self::PREC { value: _, content }
            | Self::PREC_LEFT { value: _, content }
            | Self::PREC_RIGHT { value: _, content }
            | Self::PREC_DYNAMIC { value: _, content } => content.as_seq(),
            _ => None,
        }
    }
}

#[derive(Clone, Debug, PartialEq, Eq, Deserialize, Serialize)]
#[serde(untagged)]
pub enum PrecedenceValue {
    Integer(i32),
    Name(String),
}

impl From<i32> for PrecedenceValue {
    fn from(value: i32) -> Self {
        Self::Integer(value)
    }
}

impl Grammar {
    /// Starting from `rule_name`, find all symbols (named or anonymous) which can be reached.
    pub fn reachable_set<'a>(&'a self, rule_name: &str) -> Option<HashSet<&'a str>> {
        let mut set = HashSet::new();
        let (name, rule) = self.rules.get_key_value(rule_name)?;
        set.insert(name.as_str());
        self.compute_reachable(rule, &mut set)?;
        Some(set)
    }

    fn compute_reachable<'a>(
        &'a self,
        rule: &'a RuleDef,
        set: &mut HashSet<&'a str>,
    ) -> Option<()> {
        match rule {
            RuleDef::ALIAS {
                content,
                named: _,
                value,
            } => {
                if set.insert(value) {
                    self.compute_reachable(content, set)?;
                }
            }
            RuleDef::BLANK => {}
            RuleDef::STRING { value } => {
                set.insert(value.as_str());
            }
            RuleDef::PATTERN { value: _, flags: _ } => {}
            RuleDef::SYMBOL { name } => {
                // Don't repeat if we have already seen it before.
                if set.insert(name.as_str()) {
                    let rule = self.rules.get(name)?;
                    self.compute_reachable(rule, set)?;
                }
            }
            RuleDef::CHOICE { members } => {
                for member in members {
                    self.compute_reachable(member, set)?;
                }
            }
            RuleDef::FIELD { name: _, content } => self.compute_reachable(content, set)?,
            RuleDef::SEQ { members } => {
                for member in members {
                    self.compute_reachable(member, set)?;
                }
            }
            RuleDef::REPEAT { content } => self.compute_reachable(content, set)?,
            RuleDef::REPEAT1 { content } => self.compute_reachable(content, set)?,
            RuleDef::PREC_DYNAMIC { value: _, content } => self.compute_reachable(content, set)?,
            RuleDef::PREC_LEFT { value: _, content } => self.compute_reachable(content, set)?,
            RuleDef::PREC_RIGHT { value: _, content } => self.compute_reachable(content, set)?,
            RuleDef::PREC { value: _, content } => self.compute_reachable(content, set)?,
            RuleDef::TOKEN { content } => self.compute_reachable(content, set)?,
            RuleDef::IMMEDIATE_TOKEN { content } => self.compute_reachable(content, set)?,
            RuleDef::RESERVED {
                context_name: _,
                content,
            } => self.compute_reachable(content, set)?,
        }

        Some(())
    }
}
