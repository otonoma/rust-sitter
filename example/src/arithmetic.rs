pub mod grammar {
    use rust_sitter::Rule;
    #[derive(PartialEq, Eq, Debug, Rule)]
    #[language]
    #[extras(
        // whitespace
        re(r"\s")
    )]
    pub enum Expression {
        Number(#[leaf(pattern(r"\d+"))] i32),
        #[prec_left(1)]
        Sub(Box<Expression>, #[leaf("-")] (), Box<Expression>),
        #[prec_left(2)]
        Mul(Box<Expression>, #[leaf("*")] (), Box<Expression>),
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use grammar::Expression;
    use rust_sitter::Language;

    #[wasm_bindgen_test::wasm_bindgen_test]
    #[test]
    fn successful_parses() {
        assert_eq!(
            grammar::Expression::parse("1").into_result().unwrap(),
            Expression::Number(1)
        );

        assert_eq!(
            grammar::Expression::parse(" 1").into_result().unwrap(),
            Expression::Number(1)
        );

        assert_eq!(
            grammar::Expression::parse("1 - 2").into_result().unwrap(),
            Expression::Sub(
                Box::new(Expression::Number(1)),
                (),
                Box::new(Expression::Number(2))
            )
        );

        assert_eq!(
            grammar::Expression::parse("1 - 2 - 3").into_result().unwrap(),
            Expression::Sub(
                Box::new(Expression::Sub(
                    Box::new(Expression::Number(1)),
                    (),
                    Box::new(Expression::Number(2))
                )),
                (),
                Box::new(Expression::Number(3))
            )
        );

        assert_eq!(
            grammar::Expression::parse("1 - 2 * 3").into_result().unwrap(),
            Expression::Sub(
                Box::new(Expression::Number(1)),
                (),
                Box::new(Expression::Mul(
                    Box::new(Expression::Number(2)),
                    (),
                    Box::new(Expression::Number(3))
                ))
            )
        );

        assert_eq!(
            grammar::Expression::parse("1 * 2 * 3").into_result().unwrap(),
            Expression::Mul(
                Box::new(Expression::Mul(
                    Box::new(Expression::Number(1)),
                    (),
                    Box::new(Expression::Number(2))
                )),
                (),
                Box::new(Expression::Number(3))
            )
        );

        assert_eq!(
            grammar::Expression::parse("1 * 2 - 3").into_result().unwrap(),
            Expression::Sub(
                Box::new(Expression::Mul(
                    Box::new(Expression::Number(1)),
                    (),
                    Box::new(Expression::Number(2))
                )),
                (),
                Box::new(Expression::Number(3))
            )
        );
    }

    #[test]
    fn failed_parses() {
        insta::assert_debug_snapshot!(grammar::Expression::parse("1 + 2"));
        insta::assert_debug_snapshot!(grammar::Expression::parse("1 - 2 -"));
        insta::assert_debug_snapshot!(grammar::Expression::parse("a1"));
        insta::assert_debug_snapshot!(grammar::Expression::parse("1a"));
    }
}
