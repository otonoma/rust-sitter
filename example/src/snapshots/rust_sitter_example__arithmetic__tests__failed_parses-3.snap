---
source: example/src/arithmetic.rs
expression: "grammar::Expression::parse(\"a1\")"
---
Err(
    [
        ParseError {
            reason: FailedNode(
                [
                    ParseError {
                        reason: UnexpectedToken(
                            "a",
                        ),
                        start_byte: 0,
                        end_byte: 1,
                        start_point: Point {
                            line: 1,
                            column: 1,
                        },
                        end_point: Point {
                            line: 1,
                            column: 2,
                        },
                        text: "a",
                        kind: "ERROR",
                        parent_context: Some(
                            ParentContext {
                                kind: "ERROR",
                            },
                        ),
                    },
                ],
            ),
            start_byte: 0,
            end_byte: 1,
            start_point: Point {
                line: 1,
                column: 1,
            },
            end_point: Point {
                line: 1,
                column: 2,
            },
            text: "a",
            kind: "ERROR",
            parent_context: Some(
                ParentContext {
                    kind: "source_file",
                },
            ),
        },
    ],
)
